File: cpp/stacks/parser_example.cpp
#include <iostream>
#include <stack>

bool is_blanaced(std::string input_str){

    std::stack<char> stk;
    // std::string open_brackets[3] = {"[", "{", "("};
    // std::string close_brackets[3] = {"]", "}", ")"};

    for (char charac : input_str){
        if (charac == '[' || charac == '{' || charac =='('){
            stk.push(charac);
        } else if (charac == ']' || charac == '}' || charac ==')'){
            //stk.pop();
            char top = stk.top();
            if ((charac == ']' && top != '[') ||
                (charac == ')' && top != '(') ||
                (charac == '}' && top != '{'))
                return false; 
            else{
                stk.pop();
            }
        } else{
            continue;
        }
    }
    return stk.empty();
}

int main(){

    std::string input_str= "[[)]]";
    if (is_blanaced(input_str)){
        std::cout<<"Input string is balanced."<<std::endl;
    } else{
        std::cout<<"Input string is not balanced."<<std::endl;
    }
    
    return 0;
}

File: python/stack/parser_example.py
from collections import deque

def is_balanced(input_str : str) -> bool:

    stack = deque()
    bracket_pairs = { '(' : ')',
                      '[' : ']',
                      '{' : '}', }

    for char in input_str:
        if char in bracket_pairs.keys():
            stack.append(char)
        elif char in bracket_pairs.values():
            if stack:
                top = stack.pop()
                if bracket_pairs[top] != char:
                    return False
            else:
                return False
        else:
            # char is not bracket, so we ignore it.
            continue
    return not stack

if __name__ == "__main__":

    input_str = "{([[]])}"
    
    if is_balanced(input_str):
        print("Input string is balanced")
    else:
        print("Input string is not balanced.")